sympy/core/expr.py

```diff
@@ -752,6 +749,12 @@ def equals(self, other, failing_expression=False):
             # then this can't be a zero
             return False
 
+        factors = diff.as_coeff_mul()[1]
+        if len(factors) > 1:  # avoid infinity recursion
+            fac_zero = [fac.equals(0) for fac in factors]
+            if None not in fac_zero:  # every part can be decided
+                return any(fac_zero)
```

Comment:
Apart from the type hints is this the main change in this PR?

Reply:
Note that mixing changes together like this makes the diff much more difficult to review because the important behavioural changes are buried in a sea of unrelated changes.

Reply:
No the main change is in `evalf_pow`.
Sorry about mixing changes together. Originally I only want to fix the `0**-1` evalf bug, then I feel the `evalf.py` is hard to modify without type hints, then the fix makes some tests fail and I have to patch other modules, then `mypy` reports tons of errors...
Since CI passes now, I will re-arrange the commits by logic.